### PR TITLE
initial commit of developer faq and best practices

### PR DESCRIPTION
**Pull request by jsong785**

### CODE REVIEW

File: docs/references/faq.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+## _"Internal error: Failed to read port assignment"_
+
+When you encounter a "…Failed to read port assignment…" (below) error message, the cause is almost always a mismatch of versions. That is, you use a V1 plugin in Designer version 2021.4+ that requires a V2 plugin.
+
+![Use Port](assets/failed_to_read_port.png)
+
+Another cause might be incorrect plugin code. You can diagnose the issue [here](#can-i-run-plugins-outside-of-designer).
+
+## _"How can I update tool config?" Or, "update_tool_config" is not saving"_
```
Comment (ayx-orion) [inline docs/references/faq.md]: mismatched quotes here

**Comment by ayx-orion** [comment]
LGTM other than the quotes, will let david D handle grammer/wording/etc

**Review by bek44 — APPROVED** [review]

**Review by paul-england — APPROVED** [review]

File: docs/references/best_practices.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+Go to [The Fastest Way to Read a CSV in Pandas](https://pythonspeed.com/articles/pandas-read-csv-fast/).
+
+## Multiprocessing/Threading
+
+You should think carefully before you attempt this, as it introduces complexity. Furthermore, the AMP engine maintains a thread pool of its own—it attempts to keep those threads at maximum utilization.
+
+If you're considering a second process/thread, a better option is to write a new tool and chain it together in the workflow. The AMP engine thread pool scheduler attempts to keep various plugins running as necessary. Furthermore, certain languages are only concurrent, not parallel. If you use Python, CPython is likely in use and has the Global Interpreter Lock (GIL).
+
+If you intend to deplay an "all-in-one" tool, Designer offers [Macro](https://help.alteryx.com/20223/designer/macros) functionality.
```
Comment (ayx-david) [inline docs/references/best_practices.md:49]: @jsong785 might be a typo - assuming "deplay" should be "deploy"?

File: docs/references/faq.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+If you encounter this error (as shown below) _"This tool is only supported in the AMP Engine. Please enable AMP to use this tool."_:
+
+![Error (use AMP)](assets/tool_is_only_supported_in_amp.png)
+ 
+Check that the AMP engine is enabled. In Designer's Workflow Configuration window, select the Runtime tab and ensure the Use AMP Engine check box is checked.
+
+![Toggle AMP](assets/use_amp.png)
+ 
+If you still encounter the error with the correct configuration, make sure you are using a plugin produced with the SDK V2 and up.
```
Comment (ayx-david) [inline docs/references/faq.md:24]: @jsong785 Update to:

If you still encounter the error with the correct configuration, make sure the plugin was produced with version 2+ of the SDK.

File: docs/references/faq.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+:information_source: The version of Python SDK that the tool will be using can be found at `requirements-thirdparty.txt`.
```
Comment (ayx-david) [inline docs/references/faq.md:26]: @jsong785 Update to:

The version of Python SDK that the tool uses is located in `requirements-thirdparty.txt`.

File: docs/references/best_practices.md (hunk 2 above)
Comment (jsong785) [reply docs/references/best_practices.md:49]: Handled in https://github.com/alteryx/ayx-developer-sdk/pull/6

File: docs/references/faq.md (hunk 3 above)
Comment (jsong785) [reply docs/references/faq.md:24]: Handled in https://github.com/alteryx/ayx-developer-sdk/pull/6

File: docs/references/faq.md (hunk 4 above)
Comment (jsong785) [reply docs/references/faq.md:26]: Handled in https://github.com/alteryx/ayx-developer-sdk/pull/6

**Pull request merged**